machinery/objects.go

```diff
@@ -73,7 +80,84 @@ type objectEngineDivergeDetector interface {
 	) (res DivergeResult, err error)
 }
 
+// Teardown ensures the given object is safely removed from the cluster.
+func (e *ObjectEngine) Teardown(
+	ctx context.Context,
+	owner client.Object, // Owner of the object.
+	revision int64, // Revision number, must start at 1.
+	desiredObject *unstructured.Unstructured,
+) (objectDeleted bool, err error) {
+	// Sanity checks.
+	if revision == 0 {
+		panic("owner revision must be set and start at 1")
+	}
+	if len(owner.GetUID()) == 0 {
+		panic("owner must be persistet to cluster, empty UID")
```

Comment:
typo in: `persistet`
speaking of it, a naive question: 
I think this is intended to be called by implementor from an operator/3rd application, so why do we _panic_ causing whole program crash instead of returning the err and let the implementor handle them according to their needs?

Reply:
```suggestion
		panic("owner must be persisted to cluster, empty UID")
```

Reply:
In most cases the application does not crash, as almost all long-running or request serving Go applications should recover from panics.
I've opted to panic in these cases, because I consider these errors a developer error.
These kinds of errors should never go unnoticed and I think it's warranted to potentially crash a users application if they are using the library wrong,